SciSpaCy/custom_sentence_segmenter.py

```diff
@@ -0,0 +1,91 @@
+def combined_rule_sentence_segmenter(doc):
+    # keep track of the two previous tokens
+    prev_tokens = [None, None]
+
+    # keep stacks for determining when we are inside parenthesis or brackets
+    open_parens = []
+    open_brackets = []
+    for token in doc:
+        # handling special quote symbols
+        if token.text == '“' or token.text == '”':
+            if prev_tokens[-1] and prev_tokens[-1].text != ".":
+                doc[token.i].is_sent_start = False
+
+        if token.text[0].isdigit():
+            # handling an abbrevation followed by a number
+            abbreviations = ["sec.",
+                             "secs.",
+                             "Sec.",
+                             "Secs.",
+                             "fig.",
+                             "figs.",
+                             "Fig.",
+                             "Figs.",
+                             "eq.",
+                             "eqs.",
+                             "Eq.",
+                             "Eqs.",
+                             "no.",
+                             "nos.",
+                             "No.",
+                             "Nos."]
+            if prev_tokens[-1] and prev_tokens[-1].text in abbreviations:
+                doc[token.i].is_sent_start = False
+
+            # handle a bracket followed by a number
+            if prev_tokens[-1] and prev_tokens[-1].text == '[':
+                doc[token.i].is_sent_start = False
+
+
+        # sentences can only start with ( if there is a complete sentence within the parens
+        # here a . is serving as a proxy for being a complete sentence
+        # This isn't quite correct, as a sentence may start with (Bottom left) and this prevents that
+        # from being segmented correctly
+        if token.text == "(":
+            open_parens.append(token)
+        if token.text == ")" and open_parens != []:
+            last_open_paren = open_parens.pop()
+            if prev_tokens[-1] and prev_tokens[-1].text != ".":
+                # allow things like (A) to start a sentence
+                if not (last_open_paren.i == (token.i-2) and len(prev_tokens[-1].text) == 1):
+                    doc[last_open_paren.i].is_sent_start = False
+
+        # same logic as above but for brackets instead of parens
+        if token.text == "[":
```

Comment:
It would be great if this if statement is combined with the one above instead of copying the code.

Reply:
👍